[RFC] Make the HOME environment variable optional

JeromeDeBretagne opened the pull request:
Stop the assumption that every operating system or setup has a HOME directory
available (or a similar notion) with a corresponding environment variable set.
This assumption is not always accurate as raised in ERL-476, for example on
Android with its stricter model of sandboxed applications.


As suggested in https://bugs.erlang.org/browse/ERL-476 here is a first pull
request to get comments on whether this is going in a right direction or not.

I've found 18 occurrences of init:get_argument(home). There seems to
be very few cases where it is actually required :
- in dialyzert_plt.erl on line 239, but the DIALYZERL_PLT env can bet set
- in dbg_wx_settings.erl on line 77
- in observer_wx.erl on line 647

In practice, Dialyzer and graphical user interfaces won't be run in the
scenarios that ERL-476 is trying to cover and support. 

In other parts of the code, there is already a fallback in place:
- net_adm.erl on line 45, with the option to use the current directory
- ssh_file.erl on line 388, which is to pass a user_dir

In the remaining parts finally, the home arg usage is in an optional feature, 
is checked already to return an error when not set, or is used in some test
cases that won't be run in the scenarios covered by ERL-476. 

I've checked all the getenv("HOME") calls too and I've found the same patterns.

I'm getting the impression that the HOME environment can already be made
optional as-is, with the small change from this pull request.

Any inputs, ideas or suggestions? Thanks a lot.


P.S. I've fixed a few typos and removed many trailing spaces, I can split this
pull request in two commits if preferred.

bjorng commented:
> P.S. I've fixed a few typos and removed many trailing spaces, I can split this
> pull request in two commits if preferred.

Yes, please split up this commit.

JeromeDeBretagne commented:
Here you go, I've just split up into 2 commits as suggested.
I've also updated the copyright end date as I had forgotten to do so at first.

JeromeDeBretagne commented:
I've rebased this branch to remove a merge conflict. I've also split the change removing the trailing whitespaces from the change fixing some comment typos.

If you don't want to merge the whitespace part, let me know and I'll simply drop the last commit. 

Thanks,
Jérôme

JeromeDeBretagne commented:
👍 @garazdawi Thanks Lukas !
